src/app/SimpleSubscriptionResumptionStorage.cpp

```diff
@@ -0,0 +1,461 @@
+    {
+        SubscriptionInfo currentSubscriptionInfo;
+        CHIP_ERROR err = Load(subscriptionIndex, currentSubscriptionInfo);
+
+        // if empty and firstEmptySubscriptionIndex isn't set yet, then mark empty spot
+        if ((firstEmptySubscriptionIndex == CHIP_IM_MAX_NUM_SUBSCRIPTIONS) && (err == CHIP_ERROR_PERSISTED_STORAGE_VALUE_NOT_FOUND))
+        {
+            firstEmptySubscriptionIndex = subscriptionIndex;
+        }
+
+        // delete duplicate
+        if (err == CHIP_NO_ERROR)
+        {
+            if ((subscriptionInfo.mNodeId == currentSubscriptionInfo.mNodeId) &&
+                (subscriptionInfo.mFabricIndex == currentSubscriptionInfo.mFabricIndex) &&
+                (subscriptionInfo.mSubscriptionId == currentSubscriptionInfo.mSubscriptionId))
+            {
+                Delete(subscriptionIndex);
+                // if duplicate is the first empty spot, then also set it
+                if (firstEmptySubscriptionIndex == CHIP_IM_MAX_NUM_SUBSCRIPTIONS)
+                {
+                    firstEmptySubscriptionIndex = subscriptionIndex;
+                }
+            }
+        }
+    }
+
+    // Fail if no empty space
+    if (firstEmptySubscriptionIndex == CHIP_IM_MAX_NUM_SUBSCRIPTIONS)
+    {
+        return CHIP_ERROR_NO_MEMORY;
+    }
+
+    // Now construct subscription state and save
+    Platform::ScopedMemoryBuffer<uint8_t> backingBuffer;
+    backingBuffer.Calloc(MaxSubscriptionSize());
+    ReturnErrorCodeIf(backingBuffer.Get() == nullptr, CHIP_ERROR_NO_MEMORY);
+
+    TLV::ScopedBufferTLVWriter writer(std::move(backingBuffer), MaxSubscriptionSize());
+
+    ReturnErrorOnFailure(Save(writer, subscriptionInfo));
+
+    const auto len = writer.GetLengthWritten();
+    VerifyOrReturnError(CanCastTo<uint16_t>(len), CHIP_ERROR_BUFFER_TOO_SMALL);
+
+    writer.Finalize(backingBuffer);
+
+    ReturnErrorOnFailure(
+        mStorage->SyncSetKeyValue(DefaultStorageKeyAllocator::SubscriptionResumption(firstEmptySubscriptionIndex).KeyName(),
+                                  backingBuffer.Get(), static_cast<uint16_t>(len)));
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR SimpleSubscriptionResumptionStorage::Delete(NodeId nodeId, FabricIndex fabricIndex, SubscriptionId subscriptionId)
+{
+    CHIP_ERROR deleteErr = CHIP_ERROR_PERSISTED_STORAGE_VALUE_NOT_FOUND;
+
+    uint16_t count = 0;
+    for (uint16_t subscriptionIndex = 0; subscriptionIndex < CHIP_IM_MAX_NUM_SUBSCRIPTIONS; subscriptionIndex++)
+    {
+        SubscriptionInfo subscriptionInfo;
+        CHIP_ERROR err = Load(subscriptionIndex, subscriptionInfo);
+
+        // delete duplicate
+        if (err == CHIP_NO_ERROR)
+        {
+            if ((nodeId == subscriptionInfo.mNodeId) && (fabricIndex == subscriptionInfo.mFabricIndex) &&
+                (subscriptionId == subscriptionInfo.mSubscriptionId))
+            {
+                deleteErr = Delete(subscriptionIndex);
```

Comment:
What if I:
  - (a) have a duplicate somehow
  - (b) deleting the first duplicate fails, but the 2nd deletion succeeds
  
  We still return success (last delete error code wins) even though we had a failure.
  
  Should we early-exit here on failure? Or at least return the 'most error restult"?

Reply:
I've made changes to return the last delete error if one exists.